cirq-core/cirq/sim/simulator_base.py

```diff
@@ -179,13 +207,43 @@ def _core_iterator(
                             continue
                         if self._ignore_measurement_results:
                             op = ops.phase_damp(1).on(*op.qubits)
-                    sim_state.axes = tuple(sim_state.qubit_map[qubit] for qubit in op.qubits)
-                    protocols.act_on(op, sim_state)
+
+                    # Go through the op's qubits and join any disparate ActOnArgs states
+                    # into a new combined state.
+                    op_args: Optional[TActOnArgs] = None
+                    for q in op.qubits if len(op.qubits) != 0 else qubits:
+                        if op_args is None or q not in op_args.qubits:
+                            op_args = (
+                                sim_state[q]
+                                if op_args is None
+                                else cast(TActOnArgs, op_args).join(sim_state[q])
```

Comment:
I don't think this `cast` is necessary - mypy should see the `if/else` and recognize that `op_args` must be a `TActOnArgs` object.